### PR TITLE
17903 Upgrade to Vue-CLI v5 + upgrade ESLint, etc + updates for Node 20

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* bcgov/entity#17903

*Description of changes:*
- app version = 5.2.0
- updated Dockerfile
- updated CD script
- updated CI script
- updated PR script
- added ESLint ignore file
- updated ESLint config
- updated Vue CLI to v5
- updated misc other dependencies (eslint, TS, etc)
- auto-fixed some code (let -> const)
- overrode misc eslint problems
- auto-fixed template attributes
- fixed some property/method visibilities
- fixed App unit test (incomplete store)
- fixed type error (object not iterable)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namerequest license (Apache 2.0).

### CODE REVIEW

**Review by JazzarKarim — APPROVED** [review]

**Comment by severinbeauvais** [comment]
@pwei1018 Can you help me find why `npm run build` is failing, please?

**Comment by severinbeauvais** [comment]
> @pwei1018 Can you help me find why `npm run build` is failing, please?

I will try upgrading to Vue-CLI v5. If that doesn't work then maybe we should consider changing to Vite. Thoughts?

File: src/App.vue [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
-                 :webChatUrl = "window['webChatUrl']"
-                 :webChatStatusUrl = "window['webChatStatusUrl']"
+      <ChatPopup
+        v-if="enableOldWebchat"
+        :openTooltipMessage="openTooltipMessage"
+        :axios="axios"
+        :isMobile="isMobile"
+        :webChatReason="window['webChatReason']"
+        :webChatUrl="window['webChatUrl']"
+        :webChatStatusUrl="window['webChatStatusUrl']"
```
Comment (severinbeauvais) [inline src/App.vue:11]: This, and most changes in the vue files below, was auto-fixed by eslint.

File: src/components/common/applicant-info-1.vue [hunk 2]
```diff
@@ -1,133 +1,190 @@
...
                 offset-y
-                v-model="showAddressMenu"
               >
-                <!-- Using activator slot to avoid having to absoluetly position v-menu. Visibility is controlled by
-                "showAddressMenu" field, "on" function provided by slot is used only to supress error message-->
-                <template v-slot:activator="{ on }">
+                <!-- Using activator slot to avoid having to absolutely position v-menu. Visibility is controlled by
+                "showAddressMenu" field, "on" function provided by slot is used only to supress error message. -->
+                <!-- eslint-disable-next-line vue/no-unused-vars -->
+                <template #activator="{ on }">
```
Comment (severinbeauvais) [inline src/components/common/applicant-info-1.vue:126]: "on" is not used here (weird workaround per comment above) so I simply ignored the eslint problem here.

I will test this before merging.

File: src/components/common/request-details.vue [hunk 3]
```diff
@@ -3,59 +3,77 @@
...
+          <li v-if="nameChoices.name3">
+            3. {{ nameChoices.name3 }} {{ nameChoices.designation3 }}
+          </li>
         </template>
       </ul>
     </section>
 
     <v-row>
-      <v-col col="6">
+      <v-col class="col">
```
Comment (severinbeauvais) [inline src/components/common/request-details.vue:24]: Hmm, weird auto-fix.

I will test this before merging.

File: src/components/common/request-details.vue [hunk 4]
```diff
@@ -3,59 +3,77 @@
...
-            {{`${applicant.countryTypeCd === Location.CA ? 'Canada' : applicant.countryTypeCd},
-             ${applicant.postalCd ? applicant.postalCd : ''}`}}
+            {{ `${applicant.countryTypeCd === Location.CA ? 'Canada' : applicant.countryTypeCd},
+             ${applicant.postalCd ? applicant.postalCd : ''}` }}
           </li>
         </ul>
       </v-col>
 
-      <v-col col="6">
+      <v-col class="col">
```
Comment (severinbeauvais) [inline src/components/common/request-details.vue:42]: I will test this before merging.

File: src/components/new-request/business-fetch.vue [hunk 5]
```diff
@@ -28,12 +31,7 @@ import Vue from 'vue'
 import { Component, Emit } from 'vue-property-decorator'
 import { Action } from 'vuex-class'
 import { BusinessSearchIF } from '@/interfaces'
-
-enum States {
-  INITIAL = 'initial',
-  SEARCHING = 'searching',
-  SUMMARY = 'summary'
-}
```
Comment (severinbeauvais) [inline src/components/new-request/business-fetch.vue:36]: TS complained about this so I moved it into a new file in enums folder.

File: src/components/new-request/business-lookup.vue [hunk 6]
```diff
@@ -1,28 +1,31 @@
 <template>
   <!-- once in Summary state, need to re-mount to reuse this component -->
-  <div id="business-lookup" v-if="state !== States.SUMMARY">
+  <div
+    v-if="state !== States.SUMMARY"
+    id="business-lookup"
+  >
     <v-autocomplete
+      v-model="searchField"
```
Comment (severinbeauvais) [inline src/components/new-request/business-lookup.vue:8]: ESLint changed this and then I fixed it manually as per [similar code here](https://github.com/bcgov/bcrs-shared-components/blob/14b5f1e234826a04fb619588acbc05f058b935f6/src/components/business-lookup/BusinessLookup.vue#L5).

I will test this before merging.

File: src/components/new-request/name-check/name-check-conflicts.vue [hunk 7]
```diff
@@ -191,23 +364,22 @@ import { NameCheckErrorType, NameCheckItemType } from '@/enums'
...
-  private expanded: Array<NameCheckItemIF> = []
-  private headers = [
+  expanded: Array<NameCheckItemIF> = []
+  headers = [
     { text: 'Icon', value: 'icon' },
     { text: 'Problem', value: 'problem' },
     { text: 'Words', value: 'words' },
     { text: '', value: 'data-table-expand' }
   ]
-  private readonly NameCheckItemType = NameCheckItemType
```
Comment (severinbeauvais) [inline src/components/new-request/name-check/name-check-conflicts.vue:210]: A few visibility fixes here (as I was looking at other problems in this file).

File: src/components/new-request/submit-request/reserve-submit.vue [hunk 8]
```diff
@@ -125,18 +131,16 @@ export default class ReserveSubmit extends Vue {
 
     let data: any
     let request: NameRequestI
-    switch (setup) {
+    switch (setup as any) {
```
Comment (severinbeauvais) [inline src/components/new-request/submit-request/reserve-submit.vue:134]: ESLint didn't like the previous "@ts-ignore" comments so I worked around it this way.

File: src/interfaces/models.ts [hunk 9]
```diff
@@ -11,12 +11,6 @@ import {
 
 export type SubmissionTypeT = 'examination' | 'consent' | 'normal'
 
-export interface AnalysisJSONI {
-  header?: string
-  issues: IssueI[]
-  status: 'fa' | 'rc' | 'ar'
-}
-
```
Comment (severinbeauvais) [inline src/interfaces/models.ts:19]: I moved this below the declaration for `IssueI`.

Ditto for other changes in here.

File: src/interfaces/new-request-interface.ts [hunk 10]
```diff
@@ -4,7 +4,7 @@ import { BusinessSearchIF, NameChoicesIF, NrDataIF, RequestOrConsentIF } from '@
 import { CompanyTypes, EntityTypes, Location, NrAffiliationErrors, NrRequestActionCodes, NrRequestTypeCodes }
   from '@/enums'
 
-interface RequestNameMapI extends RequestNameI {}
+type RequestNameMapI = RequestNameI
```
Comment (severinbeauvais) [inline src/interfaces/new-request-interface.ts:7]: Auto-fix. Pretty smart!

File: src/store/actions.ts [hunk 11]
```diff
@@ -251,20 +251,21 @@ const commitExistingData = ({ commit, getters }): void => {
...
   if (!getters.getEntityTypeOptions.some(option => option.value === entity_type_cd)) {
-    let obj = getters.getEntityTypesBC.find(entity => entity.value === entity_type_cd)
+    const obj = getters.getEntityTypesBC.find(entity => entity.value === entity_type_cd)
       ? getters.getEntityTypesBC.find(entity => entity.value === entity_type_cd)
       : getters.getEntityTypesXPRO.find(entity => entity.value === entity_type_cd)
     commit('mutateEntityTypeAddToSelect', obj)
   }
-  let { requestTypeCd, request_action_cd } = getters.getNr
+  const { requestTypeCd } = getters.getNr
+  let { request_action_cd } = getters.getNr
```
Comment (severinbeauvais) [inline src/store/actions.ts:260]: ESLint complained about `requestTypeCd` needing to be a const, but `request_action_cd` cannot be const, so I broke up this declaration like this.

File: Makefile [hunk 12]
```diff
@@ -25,7 +25,7 @@ lint:  ## Run linting ofcode.
 	npm run lint
 
 test:  ## Unit testing
-	npm run test
+	npm run test:unit
```
Comment (severinbeauvais) [inline Makefile:28]: Same as Create UI, Edit UI and Filings UI.

File: src/App.vue [hunk 13]
```diff
@@ -280,6 +289,9 @@ export default class App extends Mixins(
 
   /** Fetches and stores the current JS date. */
   private async updateCurrentJsDate (): Promise<void> {
+    // don't run in Jest tests
+    if (process.env.JEST_WORKER_ID) return
```
Comment (severinbeauvais) [inline src/App.vue:293]: To get rid of a console.log during unit tests.

File: src/mixins/update-user-mixin.ts [hunk 14]
```diff
@@ -6,6 +6,9 @@ import AuthServices from '@/services/auth-services'
 export class UpdateUserMixin extends Vue {
   /** Fetches the user's info and updates LaunchDarkly. */
   async updateUser (): Promise<any> {
+    // don't run in Jest tests
+    if (process.env.JEST_WORKER_ID) return
```
Comment (severinbeauvais) [inline src/mixins/update-user-mixin.ts:10]: To get rid of a multiple "console.info" during unit tests.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-710-uutlgp48.web.app

File: src/components/common/applicant-info-1.vue (hunk 2 above)
Comment (severinbeauvais) [reply src/components/common/applicant-info-1.vue:126]: Works fine.

image

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-710-uutlgp48.web.app

File: src/components/new-request/business-fetch.vue (hunk 5 above)
Comment (JazzarKarim) [reply src/components/new-request/business-fetch.vue:36]: Out of curiosity, what was TS complaining about exactly Sev? That enums have to be separate and not be defined in the components?

Comment (severinbeauvais) [reply src/components/new-request/business-fetch.vue:36]: It said they were "unused", even though I was clearly using them 🤷‍♂️

**Comment by severinbeauvais** [comment]
Update: so this builds cleanly and the unit tests are happy (the few we have), but I'm experiencing this error while testing:

image

Thus, don't merge this until I have it all sorted out.

File: tests/unit/App.spec.ts [hunk 15]
```diff
@@ -14,14 +12,11 @@ import {
...
+import { getVuexStore } from '@/store/'
 
 const vuetify = new Vuetify({})
 
-const store = new Vuex.Store<any>({
-  getters: {
-    isMobile (): any {}
-  }
-})
+const store = getVuexStore()
```
Comment (severinbeauvais) [inline tests/unit/App.spec.ts:19]: This is the fix for "isRoleStaff is not defined" and other getter/action test issues.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-710-uutlgp48.web.app

File: src/mixins/payment-mixin.ts [hunk 16]
```diff
@@ -71,8 +71,9 @@ export class PaymentMixin extends Mixins(ActionMixin) {
     return this.$store.getters[paymentTypes.GET_PAYMENT_RECEIPT]
   }
 
-  get paymentFees () {
-    return this.$store.getters[paymentTypes.GET_PAYMENT_FEES]
+  get paymentFees (): any[] {
+    const fees = this.$store.getters[paymentTypes.GET_PAYMENT_FEES]
+    return [fees]
```
Comment (severinbeauvais) [inline src/mixins/payment-mixin.ts:76]: This is the fix for "paymentFees is not iterable".

The root cause is that several Vue components pass `paymentFees` to FeeSummary.vue, which expects an array of fee objects. However, `fees` is [stored as an object, not an array](https://github.com/bcgov/namerequest/blob/c8e8b6722adca77bccaca97d18c934e8506a2045/src/modules/payment/store/index.ts#L36). So, this code change now provides "paymentFees" as an array of a single object, and everyone is happy.

Comment (severinbeauvais) [reply src/mixins/payment-mixin.ts:76]: I'm guessing that the previous version of Typescript handled an object as equivalent to an array of a single object, which is why this worked before.

Here it is working now:

image

File: src/components/common/request-details.vue (hunk 3 above)
Comment (severinbeauvais) [reply src/components/common/request-details.vue]: New on the left, old on the right. Works fine.

image

File: src/components/common/request-details.vue (hunk 4 above)
Comment (severinbeauvais) [reply src/components/common/request-details.vue]: Same as above. (Works fine.)

Note that `class="col"` is NOT the final code. This is an old snapshot.

File: src/components/new-request/business-lookup.vue (hunk 6 above)
Comment (severinbeauvais) [reply src/components/new-request/business-lookup.vue:8]: I broke this. I'm working on a fix now.

Comment (severinbeauvais) [reply src/components/new-request/business-lookup.vue]: I reverted the code change and added an ESLint rule exception instead.

I tested this and the business lookup works as before.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-710-uutlgp48.web.app

File: src/components/new-request/business-fetch.vue (hunk 5 above)
Comment (JazzarKarim) [reply src/components/new-request/business-fetch.vue:36]: Well that's weird, haha. All right Sev.

**Review by eve-git — APPROVED** [review]
Excellent!

**Pull request merged**